include/ultra64/gbi.h

```diff
@@ -1335,89 +1384,384 @@ typedef struct {
+                { r3, g3, b3 }, 0,              \
+                { x3, y3, z3 }, 0,              \
+            }},                                 \
+            {{                                  \
+                { r4, g4, b4 }, 0,              \
+                { r4, g4, b4 }, 0,              \
+                { x4, y4, z4 }, 0,              \
+            }},                                 \
+            {{                                  \
+                { r5, g5, b5 }, 0,              \
+                { r5, g5, b5 }, 0,              \
+                { x5, y5, z5 }, 0,              \
+            }},                                 \
+            {{                                  \
+                { r6, g6, b6 }, 0,              \
+                { r6, g6, b6 }, 0,              \
+                { x6, y6, z6 }, 0,              \
+            }},                                 \
+            {{                                  \
+                { r7, g7, b7 }, 0,              \
+                { r7, g7, b7 }, 0,              \
+                { x7, y7, z7 }, 0,              \
+            }}                                  \
+        }                                       \
+    }
+
+
+#define gdSPDefLookAt(rightx, righty, rightz, upx, upy, upz)    \
+    {{                                                          \
+        {{                                                      \
+            {      0,      0,      0 }, 0,                      \
+            {      0,      0,      0 }, 0,                      \
+            { rightx, righty, rightz }, 0,                      \
+        }},                                                     \
+        {{                                                      \
+            {   0, 128,   0 }, 0,                               \
+            {   0, 128,   0 }, 0,                               \
+            { upx, upy, upz }, 0,                               \
+        }},                                                     \
+    }}
+
+typedef struct {
+    int          cmd  : 8;
+    unsigned int type : 8;
+    unsigned int len  : 16;
+    union {
+        void       (*callback)();
+        const char*  str;
+        unsigned int u32;
+        float        f32;
+        void*        addr;
+    } value;
+} Gnoop;
 
 /*
  *  Graphics DMA Packet
  */
 typedef struct {
-    int     cmd:8;
-    unsigned int    par:8;
-    unsigned int    len:16;
-    unsigned int    addr;
+    int          cmd : 8;
+    unsigned int par : 8;
+    unsigned int len : 16;
+    unsigned int addr;
 } Gdma;
 
+#ifdef F3DEX_GBI_2
+typedef struct {
+    int          cmd : 8;
+    unsigned int len : 8;
+    unsigned int ofs : 8;
+    unsigned int par : 8;
+    unsigned int addr;
+} Gdma2;
+#endif
+
+/*
+ *  Graphics Moveword Packet
+ */
+typedef struct {
+    int          cmd    : 8;
+    unsigned int offset : 16;
+    unsigned int index  : 8;
+    unsigned int data;
+} Gmovewd;
```

Comment:
`#define gMoveWd(pkt, index, offset, data) gDma1p((pkt), G_MOVEWORD, data, offset, index)` (`F3DEX_GBI_2`) seems to suggest `offset` and `index` are swapped?
but for f3dex (not 2) it's accurate

I think

Reply:
You're correct it's inaccurate for F3DEX2, however for some reason they hacked around it in ucode_disas instead of changing the struct:
```c
Gdma dma = ptr->dma;
Gmovewd movewd = ptr->movewd;

movewd.index = dma.par;
movewd.offset = dma.len;
movewd.data = dma.addr;
```
So unfortunately it must be like this. I'll add a comment mentioning this.